cmd/util/cmd/epochs/cmd/templates.go

```diff
@@ -50,7 +50,7 @@ transaction(name: String,
 
             // the below arguments are unused and are safe to be left empty
             collectorClusters: [],
-            clusterQCs: [],
-            dkgPubKeys: [])
+            clusterQCs: [] as [FlowClusterQC.ClusterQC],
+            dkgPubKeys: [] as [String])
```

Comment:
isn't this needed for the preceding `collectorClusters: [],` argument also

Reply:
I'm actually more curious about how this worked 🤔  This change (type inferring changes - which I assume the cause of this) was done many moons ago

Reply:
Checking again, it seems that the change here does not affect the results of any tests. I do not know if the reason is because this code is not used in a test, or is there some other reason.

I made these changes together with making the changes in `fvm/blueprints/epochs.go`. Those definitely affect tests. Without those changes the following tests fail:
```
--- FAIL: TestFungibleTokenTracker (0.36s)
panic: failed to deploy Epoch contract: [Error Code: 1101] cadence runtime error Execution failed:
error: cannot infer type from array literal:  requires an explicit type annotation
  --> 0d705cc7942cbee81c7bb25b82977f305f6ec04f28e14bfd5924b4b33536d5a4:27:14
   |
27 | 		clusterQCs: [],
   |               ^

error: cannot infer type from array literal:  requires an explicit type annotation
  --> 0d705cc7942cbee81c7bb25b82977f305f6ec04f28e14bfd5924b4b33536d5a4:28:14
   |
28 | 		dkgPubKeys: [],
   |               ^
```

Reply:
I was wondering about these changes before, as I expected the type assertions not to be necessary.
Shouldn't type inference propagate the parameter type to the argument here?

Reply:
Thanks, @janezpodhostnik for double-checking.

@turbolent This particular case (extra arguments) is not supported yet (along with generics) https://github.com/onflow/cadence/blob/481cb23a7d6bac00083cf5127a86a17b900c99aa/runtime/sema/check_invocation_expression.go#L450-L451. This one should be easy to add, but generic is the complicated one. This was documented in the original issue, but can't seem to fund it now. I'll create an issue anyway for these particular ones.

Reply:
What is the way forward here? Do I check all cadence scripts in the `flow-go` and `flow-core-contracts` for empty arrays, dictionaries in arguments and update them?

Reply:
This is actually strange. I checked this scenario and it works in Cadence `v0.23.2` (IDK how/why it worked in the first place). But doesn't work in the master. So some recent must have triggered this. I'll have a look at this today.

Reply:
Ok, so I figured it. It was this change: https://github.com/onflow/cadence/pull/1027/files#diff-6cc33849623674ce4f600cda1e45c49e822a86b35be9bc61f4ed3ca83d8f0671R79-R92. Earlier, we used to assume empty dictionaries without context-type is to be Never type. But now we are more strict and expect the user to add an explicit annotation. I can special case the empty array/dictionary case to have the old behaviour, but I think the current behaviour is correct.

Assuming the current behaviour is correct: this should only affect empty arrays/dictionaries passed into a function with generics, or that use extra arguments (e.g: signer.contracts.add() where contract inits' args are passed, apart from the add() methods arguments). Maybe checking usages of signer.contracts.add() would be a good idea and might be sufficient.

Normal function calls should be working as expected.

Reply:
I checked all usages of `.contracts.add(` and only found this reference and the one in `fvm/blueprints/epochs.go` that use empty arrays or empty maps. I updated both usages.